internal/controllers/machinedeployment/mdutil/util.go

```diff
@@ -239,6 +258,33 @@ func SetNewMachineSetAnnotations(deployment *clusterv1.MachineDeployment, newMS
 	return annotationChanged
 }
 
+// CalculateMachineSetAnnotations calculates the annotations that should be set on the MachineSet.
+func CalculateMachineSetAnnotations(deployment *clusterv1.MachineDeployment, newRevision string, machineSet *clusterv1.MachineSet) map[string]string {
+	annotations := map[string]string{}
+	for k, v := range deployment.Annotations {
+		if skipCopyAnnotation(k) {
+			continue
+		}
+		annotations[k] = v
+	}
+	if machineSet != nil {
+		currentRevision, ok := machineSet.Annotations[clusterv1.RevisionAnnotation]
+		if ok && currentRevision != newRevision {
+			revisionHistory := machineSet.Annotations[clusterv1.RevisionHistoryAnnotation]
+			oldRevisions := strings.Split(revisionHistory, ",")
+			if oldRevisions[0] == "" {
+				annotations[clusterv1.RevisionHistoryAnnotation] = currentRevision
+			} else {
+				annotations[clusterv1.RevisionHistoryAnnotation] = strings.Join(append(oldRevisions, currentRevision), ",")
+			}
+		}
```

Comment:
This code only sets the revision-history annotation if `currentRevision != newRevision`.

This is fine for initially adding the annotation, but in the next reconcile we don't set it again which means SSA will drop revision-history.

=>  We have to make sure that we always preserve an existing revision history annotation